test/functional/feature_taproot.py

```diff
@@ -0,0 +1,1271 @@
+                        n += 1
+                    # But allow picking a bit higher still
+                    n += random.randrange(5)
+                    # Now pick dummy size *just* large enough that the overall construction passes
+                    dummylen = 0
+                    while not predict_sigops_ratio(n, dummylen):
+                        dummylen += 1
+                    scripts = [("s", fn(n, pubkey)[0])]
+                    for _ in range(merkledepth):
+                        scripts = [scripts, random.choice(PARTNER_MERKLE_FN)]
+                    tap = taproot_construct(pub1, scripts)
+                    standard = annex is None and dummylen <= 80 and len(pubkey) == 32
+                    add_spender(spenders, "tapscript/sigopsratio_%i" % fn_num, tap=tap, leaf="s", annex=annex, hashtype=hashtype, key=sec2, inputs=[getter("sign"), random_bytes(dummylen)], standard=standard, failure={"inputs": [getter("sign"), random_bytes(dummylen - 1)]}, **ERR_SIGOPS_RATIO)
+        # Future leaf versions
+    for leafver in range(0, 0x100, 2):
+        if leafver == LEAF_VERSION_TAPSCRIPT or leafver == ANNEX_TAG:
+            # Skip the defined LEAF_VERSION_TAPSCRIPT, and the ANNEX_TAG which is not usable as leaf version
+            continue
+        scripts = [
+            ("bare_c0", CScript([OP_NOP])),
+            ("bare_unkver", CScript([OP_NOP]), leafver),
+            ("return_c0", CScript([OP_RETURN])),
+            ("return_unkver", CScript([OP_RETURN]), leafver),
+            ("undecodable_c0", CScript([OP_PUSHDATA1])),
+            ("undecodable_unkver", CScript([OP_PUSHDATA1]), leafver),
+            ("bigpush_c0", CScript([random_bytes(MAX_SCRIPT_ELEMENT_SIZE+1), OP_DROP])),
+            ("bigpush_unkver", CScript([random_bytes(MAX_SCRIPT_ELEMENT_SIZE+1), OP_DROP]), leafver),
+            ("1001push_c0", CScript([OP_0] * 1001)),
+            ("1001push_unkver", CScript([OP_0] * 1001), leafver),
+        ]
+        random.shuffle(scripts)
+        tap = taproot_construct(pub1, scripts)
+        add_spender(spenders, "unkver/bare", standard=False, tap=tap, leaf="bare_unkver", failure={"leaf": "bare_c0"}, **ERR_CLEANSTACK)
+        add_spender(spenders, "unkver/return", standard=False, tap=tap, leaf="return_unkver", failure={"leaf": "return_c0"}, **ERR_OP_RETURN)
+        add_spender(spenders, "unkver/undecodable", standard=False, tap=tap, leaf="undecodable_unkver", failure={"leaf": "undecodable_c0"}, **ERR_UNDECODABLE)
+        add_spender(spenders, "unkver/bigpush", standard=False, tap=tap, leaf="bigpush_unkver", failure={"leaf": "bigpush_c0"}, **ERR_PUSH_LIMIT)
+        add_spender(spenders, "unkver/1001push", standard=False, tap=tap, leaf="1001push_unkver", failure={"leaf": "1001push_c0"}, **ERR_STACK_SIZE)
+        add_spender(spenders, "unkver/1001inputs", standard=False, tap=tap, leaf="bare_unkver", inputs=[b'' for _ in range(1001)], failure={"leaf": "bare_c0"}, **ERR_STACK_SIZE)
```

Comment:
take-or-leave-nit: just do `[b''] * 1001`

Reply:
Ah yes, I forgot you could do that.